### PR TITLE
Support Apple Accelerate and improve MKL integration

### PR DESCRIPTION
**Pull request by ChrisRackauckas**

### CODE REVIEW

File: src/appleaccelerate.jl [hunk 1]
```diff
@@ -0,0 +1,62 @@
+# For now, only use BLAS from Accelerate (that is to say, vecLib)
+global const libacc = "/System/Library/Frameworks/Accelerate.framework/Accelerate"
+global const libacc_info_plist = "/System/Library/Frameworks/Accelerate.framework/Versions/Current/Resources/Info.plist"
```
[the remark this answers is not in the document]
Comment (staticfloat) [reply src/appleaccelerate.jl:3]: This is not needed here.

File: src/appleaccelerate.jl [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    m, n = size(A)
+    lda  = max(1,stride(A, 2))
+    if isempty(ipiv)
+        ipiv = similar(A, BlasInt, min(size(A,1),size(A,2)))
+    end
+
+    ccall(("dgetrf\$NEWLAPACK\$ILP64", libacc), Cvoid,
+            (Ref{BlasInt}, Ref{BlasInt}, Ptr{Float64},
+            Ref{BlasInt}, Ptr{BlasInt}, Ptr{BlasInt}),
+            m, n, A, lda, ipiv, info)
```
Comment (ViralBShah) [inline src/appleaccelerate.jl:39]: Make ipiv Int32 and do:
```
ccall(("dgetrf_", libacc), Cvoid,
            (Ref{Int32}, Ref{Int32}, Ptr{Float64},
            Ref{Int32}, Ptr{Int32}, Ptr{Int32}),
            m, n, A, lda, ipiv, info)
```

That should just work on every macos version and on Intel as well as M-series macs.

**Comment by ViralBShah** [comment]
Would you ever pass `ipiv` to `aa_getrf`? If so, that probably comes in as an Int64 vector and probably should be downcast to an Int32 vector.

**Comment by ViralBShah** [comment]
I believe the smarter thing to do is to use NEWLAPACK symbols if available, falling back to the old one if not. Not sure if `getrf` is better in that or the same. In case the new LAPACK library has a faster LU, this is worth the effort. Some benchmarking is necessary first.

cc @vpuri3

**Comment by ChrisRackauckas** [comment]
> Would you ever pass ipiv to aa_getrf? If so, that probably comes in as an Int64 vector and probably should be downcast to an Int32 vector.

We allocate all of the caches so it's safe from that. It's setup so that all caches compile up front and repeated calls then reuse the cache now, even for the info ref.

**Comment by ChrisRackauckas** [comment]
> I believe the smarter thing to do is to use NEWLAPACK symbols if available, falling back to the old one if not. Not sure if getrf is better in that or the same. In case the new LAPACK library has a faster LU, this is worth the effort. Some benchmarking is necessary first.

If I did it correctly, it doesn't seem that big of a deal in https://github.com/SciML/LinearSolve.jl/pull/358

**Comment by ViralBShah** [comment]
That seems correctly done. Unless they explicitly multi-threaded the `getrf` call, all the performance actually just comes from the matmul. openblas does have the natively multi-threaded `getrf` but I think it doesn't have access to the fast matmuls on M-series that Accelerate does: https://github.com/xianyi/OpenBLAS/blob/develop/lapack/getrf/getrf_parallel.c

**Comment by ChrisRackauckas** [comment]
I just cached the 32-bit version so the allocation is saved anyways. So yeah, seems like it's better to just support 32-bit there.

**Pull request merged**